src/Sarif.Viewer.VisualStudio/FileWatcher/SarifFolderMonitor.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using EnvDTE;
+
+using EnvDTE80;
+
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.Sarif.Viewer.Options;
+using Microsoft.Sarif.Viewer.Services;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.Sarif.Viewer.FileWatcher
+{
+    /// <summary>
+    /// Handles loading & monitoring sarif logs under solution directory .sarif folder.
+    /// </summary>
+    internal class SarifFolderMonitor : IDisposable
+    {
+        public const string SarifFolderName = ".sarif";
+        public const string SarifFileExtensionName = ".sarif";
+
+        private readonly IFileSystem fileSystem;
+        private IFileWatcher fileWatcher;
+        private bool filesLoaded;
+        private readonly ILoadSarifLogService sarifLoadLogService;
+        private readonly ICloseSarifLogService sarifCloseLogService;
+
+        private string solutionFolder = null;
+
+        internal SarifFolderMonitor(
+            IFileSystem fs = null,
+            IFileWatcher watcher = null,
+            ILoadSarifLogService sarifLoadLogService = null,
+            ICloseSarifLogService closeSarifLogService = null)
+        {
+            this.fileSystem = fs ?? new FileSystem();
+            this.fileWatcher = watcher ?? new FileWatcher();
+            this.sarifLoadLogService = sarifLoadLogService ?? new LoadSarifLogService();
+            this.sarifCloseLogService = closeSarifLogService ?? new CloseSarifLogService();
+        }
+
+        public void Dispose()
+        {
+            this.fileWatcher?.Dispose();
+        }
+
+        internal void StartWatch(string solutionPath = null)
+        {
+            this.solutionFolder = solutionPath ?? this.GetSolutionDirectory();
+            if (string.IsNullOrEmpty(this.solutionFolder) || !fileSystem.DirectoryExists(this.solutionFolder))
+            {
+                return;
+            }
+
+            string sarifLogFolder = Path.Combine(this.solutionFolder, SarifFolderName);
+            if (!fileSystem.DirectoryExists(sarifLogFolder))
+            {
+                return;
+            }
+
+            // load existing sarif logs
+            this.LoadExistingSarifLogs(sarifLogFolder);
+
+            this.fileWatcher ??= new FileWatcher();
+            this.fileWatcher.FilePath = sarifLogFolder;
+            this.fileWatcher.Filter = Constants.SarifFileSearchPattern;
+
+            // here no need to watch for sarif file log updates
+            // because when load the sarif log file in viewer, its already monitored by viewer's file watcher
```

Comment:
// no need to watch for the sarif file log updates
// because when we load, the sarif log file in the viewer, it's already monitored by the viewer's file watcher

Reply:
corrected

---
In reply to: [602309039](https://github.com/microsoft/sarif-visualstudio-extension/pull/382#discussion_r602309039) [](ancestors = 602309039)